plutus-core/testlib/PlutusCore/Generators/PIR.hs

```diff
@@ -0,0 +1,1716 @@
+
+-- | Bind a term to a type in a generator.
+bindTmName :: Name -> Type TyName DefaultUni () -> GenTm a -> GenTm a
+bindTmName x ty = local $ \ e -> e { geTerms = Map.insert x ty (geTerms e) }
+
+-- | Bind term names
+bindTmNames :: [(Name, Type TyName DefaultUni ())] -> GenTm a -> GenTm a
+bindTmNames = flip $ foldr (uncurry bindTmName)
+
+-- | Create a fresh term name, bind it to a type, and use it in a generator.
+bindFreshTmName :: String -> Type TyName DefaultUni () -> (Name -> GenTm a) -> GenTm a
+bindFreshTmName name ty k = do
+  x <- genFreshName name
+  bindTmName x ty (k x)
+
+-- | Get the names and types of the constructors of a datatype.
+constrTypes :: Datatype TyName Name DefaultUni DefaultFun () -> [(Name, Type TyName DefaultUni ())]
+constrTypes (Datatype _ _ xs _ cs) = [ (c, abstr ty) | VarDecl _ c ty <- cs ]
+  where
+    abstr ty = foldr (\ (TyVarDecl _ x k) -> TyForall () x k) ty xs
+
+
+-- | Get the name and type of the match function for a given datatype.
+matchType :: Datatype TyName Name DefaultUni DefaultFun () -> (Name, (Type TyName DefaultUni ()))
+matchType (Datatype _ (TyVarDecl _ a _) xs m cs) = (m, matchType)
+  where
+    fvs = Set.fromList (a : [x | TyVarDecl _ x _ <- xs]) <>
+          mconcat [fvType ty | VarDecl _ _ ty <- cs]
+    pars = [TyVar () x | TyVarDecl _ x _ <- xs]
+    dtyp = foldl (TyApp ()) (TyVar () a) pars
+    matchType = abstr $ dtyp ->> TyForall () r Star (foldr ((->>) . conArg) (TyVar () r) cs)
+      where r = freshenTyName fvs $ TyName $ Name "r" (toEnum 0)
+            conArg (VarDecl _ _ ty) = setTarget ty
+            setTarget (TyFun _ a b) = TyFun () a (setTarget b)
+            setTarget _             = TyVar () r
+    abstr ty = foldr (\ (TyVarDecl _ x k) -> TyForall () x k) ty xs
+
+-- | Bind a datatype declaration in a generator.
+bindDat :: Datatype TyName Name DefaultUni DefaultFun ()
+        -> GenTm a
+        -> GenTm a
+bindDat dat@(Datatype _ (TyVarDecl _ a k) _ _ _) cont =
+  bindTyName a k $
+  local (\ e -> e { geDatas = Map.insert a dat (geDatas e) }) $
+  foldr (uncurry bindTmName) cont (matchType dat : constrTypes dat)
+
+-- | Bind a binding.
+bindBind :: Binding TyName Name DefaultUni DefaultFun ()
+         -> GenTm a
+         -> GenTm a
+bindBind (DatatypeBind _ dat)              = bindDat dat
+bindBind (TermBind _ _ (VarDecl _ x ty) _) = bindTmName x ty
+-- CODE REVIEW: Should we try to generate type bindings and all the recursive types without datatypes stuff?
+-- I don't think both datatypes and this stuff should actually show up in the same code, no?
+bindBind _                                 = error "unreachable"
+
+-- | Bind multiple bindings
+bindBinds :: Foldable f => f (Binding TyName Name DefaultUni DefaultFun ()) -> GenTm a -> GenTm a
+bindBinds = flip (foldr bindBind)
+
+-- * Generators for well-kinded types
+
+-- | Give a unique "least" (intentionally vaguely specified by "shrinking order")
```

Comment:
maybe it could do with some comments explaining how it works and all the careful stuff you did to avoid getting into a loop ;)